mesop/labs/chat.py

```diff
@@ -0,0 +1,228 @@
+  return me.Style(
+    display="flex",
+    flex_direction="column",
+    align_items=align_items,
+  )
+
+
+def _make_chat_bubble_style(role: Literal["user", "assistant"]) -> me.Style:
+  """Generates styles for chat bubble.
+
+  Args:
+    role: Chat bubble background color depends on the role
+  """
+  background = (
+    _COLOR_CHAT_BUBBLE_YOU if role == _ROLE_USER else _COLOR_CHAT_BUBBLE_BOT
+  )
+  return me.Style(
+    width="80%",
+    font_size="13px",
+    background=background,
+    border_radius="15px",
+    padding=me.Padding(
+      top="0px",
+      right="15px",
+      left="15px",
+      bottom="3px",
+    ),
+    margin=me.Margin(bottom="10px"),
+    border=me.Border(
+      left=_DEFAULT_BORDER_SIDE,
+      right=_DEFAULT_BORDER_SIDE,
+      top=_DEFAULT_BORDER_SIDE,
+      bottom=_DEFAULT_BORDER_SIDE,
+    ),
+  )
+
+
+@dataclass
+class Message:
+  """Chat message metadata."""
+
+  role: Literal["assistant", "user"] = "user"
+  username: str = ""
+  content: str = ""
+
+
+@me.stateclass
+class State:
+  input: str
+  output: list[Message]
+  in_progress: bool = False
+
+
+def on_input_update(State):
+  """Generic on text input handler that saves input to State using the given key.
+
+  This helper only works if you have one state instance. If use multiple state classes
+  with this helper, then only the last event handler will be stored. For more info, see
+  https://google.github.io/mesop/guides/troubleshooting/#avoid-using-closure-variables-in-event-handler.
+  """
+
+  def _on_update(e: me.InputEvent):
+    state = me.state(State)
+    setattr(state, e.key, e.value)
+
+  return _on_update
+
+
+def chat(
+  transform: Callable[[str], Generator[str, None, None] | str],
+  *,
+  title: str | None = None,
+  bot_username: str = _BOT_USERNAME_DEFAULT,
+):
+  """Creates a simple chat UI which takes in a prompt and returns a response to the
+  prompt.
+
+  This function creates event handlers for text input and output operations
+  using the provided function `transform` to process the input and generate the output.
+
+  Args:
+    transform: Function that takes in a prompt and returns a response to the prompt.
+    title: Headline text to display at the top of the UI.
+    bot_username: Name of your bot / assistant.
+  """
+  state = me.state(State)
+
+  def on_click(e: me.ClickEvent):
+    state = me.state(State)
+    if state.in_progress or not state.input:
+      return
+
+    output = state.output
+    if output is None:
+      output = []
+    output.append(Message(role=_ROLE_USER, content=state.input))
+    state.in_progress = True
+    yield
+
+    start_time = time.time()
+    output_message = transform(state.input)
+    assistant_message = Message(role=_ROLE_ASSISTANT, username=bot_username)
+    output.append(assistant_message)
+    state.output = output
+    for content in output_message:
+      assistant_message.content += content
+      if (time.time() - start_time) >= 0.5:
```

Comment:
I'd leave a comment here on why you picked 0.5. 

fwiw: I found in practice the right number is a bit tricky (too low, and we're spending too much time on the render loop; too high and we're not giving as much feedback as we could) so we may want to make this customizable in the future.

Reply:
Yeah, I think I just copied it from one of the Mesop chat examples somewhere. I lowered it to .25 for now.